File: android/app/src/main/java/com/twintipsolutions/guthealth/data/FirestoreService.kt
package com.twintipsolutions.guthealth.data

import android.util.Base64
import com.google.firebase.Timestamp
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.Query
import com.google.firebase.functions.FirebaseFunctions
import com.google.firebase.storage.FirebaseStorage
import com.google.firebase.storage.StorageMetadata
import com.twintipsolutions.guthealth.data.models.*
import kotlinx.coroutines.tasks.await
import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Date
import java.util.Locale

class FirestoreService {
    private val auth = FirebaseAuth.getInstance()
    private val db = FirebaseFirestore.getInstance()
    private val functions = FirebaseFunctions.getInstance()
    private val storage = FirebaseStorage.getInstance()

    val currentUserId: String?
        get() = auth.currentUser?.uid

    suspend fun signInAnonymously() {
        if (auth.currentUser == null) {
            auth.signInAnonymously().await()
        }
    }

    private fun userDoc() = db.collection("users").document(
        currentUserId ?: throw IllegalStateException("Not signed in")
    )

    // --- Meals ---
    suspend fun saveMeal(meal: Meal): String {
        val ref = userDoc().collection("meals").document()
        ref.set(meal).await()
        return ref.id
    }

    suspend fun getTodayMeals(): List<Meal> {
        val startOfDay = getStartOfDay()
        val snapshot = userDoc().collection("meals")
            .whereGreaterThanOrEqualTo("createdAt", Timestamp(startOfDay))
            .orderBy("createdAt", Query.Direction.DESCENDING)
            .get()
            .await()
        return snapshot.toObjects(Meal::class.java)
    }

    suspend fun getRecentEntries(limit: Int = 5): List<RecentEntry> {
        val entries = mutableListOf<RecentEntry>()
        val startOfDay = getStartOfDay()
        val ts = Timestamp(startOfDay)

        try {
            val meals = userDoc().collection("meals")
                .whereGreaterThanOrEqualTo("createdAt", ts)
                .orderBy("createdAt", Query.Direction.DESCENDING)
                .limit(5)
                .get().await()
            meals.documents.forEach { doc ->
                val meal = doc.toObject(Meal::class.java) ?: return@forEach
                val foodNames = meal.foods.joinToString(", ") { it.name }.ifEmpty { meal.mealType }
                entries.add(
                    RecentEntry(
                        type = "meal",
                        title = meal.mealType.replaceFirstChar { it.uppercase() },
                        subtitle = foodNames,
                        timestamp = meal.createdAt
                    )
                )
            }
        } catch (_: Exception) {}

        try {
            val symptoms = userDoc().collection("symptoms")
                .whereGreaterThanOrEqualTo("createdAt", ts)
                .orderBy("createdAt", Query.Direction.DESCENDING)
                .limit(5)
                .get().await()
            symptoms.documents.forEach { doc ->
                val symptom = doc.toObject(Symptom::class.java) ?: return@forEach
                entries.add(
                    RecentEntry(
                        type = "symptom",
                        title = symptom.type.replaceFirstChar { it.uppercase() },
                        subtitle = "Severity: ${symptom.severity}/10",
                        timestamp = symptom.createdAt
                    )
                )
            }
        } catch (_: Exception) {}

        try {
            val poopLogs = userDoc().collection("poopLogs")
                .whereGreaterThanOrEqualTo("createdAt", ts)
                .orderBy("createdAt", Query.Direction.DESCENDING)
                .limit(5)
                .get().await()
            poopLogs.documents.forEach { doc ->
                val log = doc.toObject(PoopLog::class.java) ?: return@forEach
                entries.add(
                    RecentEntry(
                        type = "poop",
                        title = "Bristol Type ${log.bristolType}",
                        subtitle = "${log.color.replaceFirstChar { it.uppercase() }} - ${log.urgency}",
                        timestamp = log.createdAt
                    )
                )
            }
        } catch (_: Exception) {}

        return entries.sortedByDescending { it.timestamp?.seconds ?: 0 }.take(limit)
    }

    // --- Symptoms ---
    suspend fun saveSymptom(symptom: Symptom): String {
        val ref = userDoc().collection("symptoms").document()
        val data = mutableMapOf<String, Any?>(
            "createdAt" to symptom.createdAt,
            "type" to symptom.type,
            "severity" to symptom.severity,
            "location" to symptom.location,
            "notes" to symptom.notes
        )
        if (symptom.endedAt != null) {
            data["endedAt"] = symptom.endedAt
        }
        ref.set(data).await()
        return ref.id
    }

    suspend fun markSymptomResolved(symptomId: String) {
        userDoc().collection("symptoms").document(symptomId)
            .update("endedAt", Timestamp.now()).await()
    }

    suspend fun getTodaySymptoms(): List<Symptom> {
        val startOfDay = getStartOfDay()
        val snapshot = userDoc().collection("symptoms")
            .whereGreaterThanOrEqualTo("createdAt", Timestamp(startOfDay))
            .orderBy("createdAt", Query.Direction.DESCENDING)
            .get()
            .await()
        return snapshot.toObjects(Symptom::class.java)
    }

    // --- Poop Logs ---
    suspend fun savePoopLog(poopLog: PoopLog): String {
        val ref = userDoc().collection("poopLogs").document()
        ref.set(poopLog).await()
        return ref.id
    }

    suspend fun uploadMealPhoto(uid: String, mealId: String, imageBase64: String): String {
        val bytes = Base64.decode(imageBase64, Base64.NO_WRAP)
        val ref = storage.reference.child("users/$uid/meals/$mealId.jpg")
        val metadata = StorageMetadata.Builder()
            .setContentType("image/jpeg")
            .build()
        ref.putBytes(bytes, metadata).await()
        return ref.downloadUrl.await().toString()
    }

    suspend fun uploadPoopPhoto(uid: String, logId: String, imageBase64: String): String {
        val bytes = Base64.decode(imageBase64, Base64.NO_WRAP)
        val ref = storage.reference.child("users/$uid/poopLogs/$logId.jpg")
        val metadata = StorageMetadata.Builder()
            .setContentType("image/jpeg")
            .build()
        ref.putBytes(bytes, metadata).await()
        return ref.downloadUrl.await().toString()
    }

    suspend fun updateMealPhotoUrl(mealId: String, photoUrl: String) {
        val uid = auth.currentUser?.uid ?: return
        db.collection("users").document(uid).collection("meals").document(mealId)
            .update("photoUrl", photoUrl).await()
    }

    suspend fun updatePoopLogPhotoUrl(logId: String, photoUrl: String) {
        val uid = auth.currentUser?.uid ?: return
        db.collection("users").document(uid).collection("poopLogs").document(logId)
            .update("photoUrl", photoUrl).await()
    }

    suspend fun getTodayPoopLogs(): List<PoopLog> {
        val startOfDay = getStartOfDay()
        val snapshot = userDoc().collection("poopLogs")
            .whereGreaterThanOrEqualTo("createdAt", Timestamp(startOfDay))
            .orderBy("createdAt", Query.Direction.DESCENDING)
            .get()
            .await()
        return snapshot.toObjects(PoopLog::class.java)
    }

    // --- Daily Summary ---
    suspend fun getDailySummary(dateKey: String): DailySummary? {
        val snapshot = userDoc().collection("dailySummaries").document(dateKey).get().await()
        return snapshot.toObject(DailySummary::class.java)
    }

    fun getTodayDateKey(): String {
        val sdf = SimpleDateFormat("yyyy-MM-dd", Locale.US)
        return sdf.format(Date())
    }

    // --- Correlation Reports (new narrative model) ---
    suspend fun getCorrelationReports(limit: Int = 10): List<CorrelationReport> {
        val snapshot = userDoc().collection("correlationReports")
            .orderBy("createdAt", Query.Direction.DESCENDING)
            .limit(limit.toLong())
            .get()
            .await()
        return snapshot.documents.mapNotNull { doc ->
            val data = doc.data ?: return@mapNotNull null
            CorrelationReport(
                id = doc.id,
                createdAt = data["createdAt"] as? Timestamp,
                periodStart = data["periodStart"] as? String ?: "",
                periodEnd = data["periodEnd"] as? String ?: "",
                mealsAnalyzed = (data["mealsAnalyzed"] as? Number)?.toInt() ?: 0,
                symptomsAnalyzed = (data["symptomsAnalyzed"] as? Number)?.toInt() ?: 0,
                poopLogsAnalyzed = (data["poopLogsAnalyzed"] as? Number)?.toInt() ?: 0,
                aiReport = data["aiReport"] as? String ?: "",
                disclaimer = data["disclaimer"] as? String ?: "This is not medical advice"
            )
        }
    }

    // --- Weekly Insights ---
    suspend fun getWeeklyInsights(limit: Int = 4): List<WeeklyInsight> {
        val snapshot = userDoc().collection("weeklyInsights")
            .orderBy("generatedAt", Query.Direction.DESCENDING)
            .limit(limit.toLong())
            .get()
            .await()
        return snapshot.toObjects(WeeklyInsight::class.java)
    }

    // --- FODMAP Database ---
    suspend fun getAllFodmapFoods(): List<FodmapFood> {
        val snapshot = db.collection("fodmapDatabase")
            .get()
            .await()
        return snapshot.toObjects(FodmapFood::class.java)
    }

    suspend fun searchFodmapFoods(query: String): List<FodmapFood> {
        if (query.isBlank()) return emptyList()
        val snapshot = db.collection("fodmapDatabase")
            .orderBy("name")
            .startAt(query.lowercase())
            .endAt(query.lowercase() + "\uf8ff")
            .limit(20)
            .get()
            .await()
        return snapshot.toObjects(FodmapFood::class.java)
    }

    suspend fun getFodmapFoodsByCategory(category: String): List<FodmapFood> {
        val snapshot = db.collection("fodmapDatabase")
            .whereEqualTo("category", category)
            .limit(50)
            .get()
            .await()
        return snapshot.toObjects(FodmapFood::class.java)
    }

    // --- User Profile ---
    suspend fun getUserProfile(): UserProfile? {
        val snapshot = userDoc().get().await()
        return snapshot.toObject(UserProfile::class.java)
    }

    suspend fun updateFodmapPhase(phase: String) {
        userDoc().update(
            mapOf(
                "fodmapPhase" to phase,
                "fodmapPhaseStartDate" to Timestamp.now()
            )
        ).await()
    }

    suspend fun updateUserProfile(fields: Map<String, Any>) {
        userDoc().update(fields).await()
    }

    suspend fun deleteUserData() {
        val uid = currentUserId ?: return
        val userRef = db.collection("users").document(uid)
        val subcollections = listOf("meals", "symptoms", "poopLogs", "dailySummaries", "correlations", "correlationReports", "weeklyInsights")
        for (name in subcollections) {
            val snapshot = userRef.collection(name).get().await()
            for (doc in snapshot.documents) {
                doc.reference.delete().await()
            }
        }
        userRef.delete().await()
    }

    fun signOut() {
        auth.signOut()
    }

    suspend fun deleteAccount() {
        auth.currentUser?.delete()?.await()
    }

    // --- Cloud Functions ---
    suspend fun analyzeFoodPhoto(imageBase64: String, mealType: String? = null): Map<String, Any> {
        val data = hashMapOf<String, Any>("imageBase64" to imageBase64)
        if (mealType != null) data["mealType"] = mealType
        val result = functions.getHttpsCallable("analyzeFoodPhoto").call(data).await()
        @Suppress("UNCHECKED_CAST")
        return result.getData() as Map<String, Any>
    }

    suspend fun classifyPoopPhoto(imageBase64: String): Map<String, Any> {
        val data = hashMapOf<String, Any>("imageBase64" to imageBase64)
        val result = functions.getHttpsCallable("classifyPoopPhoto").call(data).await()
        @Suppress("UNCHECKED_CAST")
        return result.getData() as Map<String, Any>
    }

    suspend fun runCorrelationEngine(daysBack: Int = 7): Map<String, Any> {
        val data = hashMapOf<String, Any>(
            "userId" to (currentUserId ?: ""),
            "daysBack" to daysBack
        )
        val result = functions.getHttpsCallable("runCorrelationEngine").call(data).await()
        @Suppress("UNCHECKED_CAST")
        return result.getData() as Map<String, Any>
    }

    // --- History: entries for a specific date range ---
    suspend fun getMealsForDateRange(startDate: Date, endDate: Date): List<Pair<String, Meal>> {
        val snapshot = userDoc().collection("meals")
            .whereGreaterThanOrEqualTo("createdAt", Timestamp(startDate))
            .whereLessThan("createdAt", Timestamp(endDate))
            .orderBy("createdAt", Query.Direction.DESCENDING)
            .get()
            .await()
        return snapshot.documents.mapNotNull { doc ->
            val meal = doc.toObject(Meal::class.java) ?: return@mapNotNull null
            doc.id to meal
        }
    }

    suspend fun getSymptomsForDateRange(startDate: Date, endDate: Date): List<Pair<String, Symptom>> {
        val snapshot = userDoc().collection("symptoms")
            .whereGreaterThanOrEqualTo("createdAt", Timestamp(startDate))
            .whereLessThan("createdAt", Timestamp(endDate))
            .orderBy("createdAt", Query.Direction.DESCENDING)
            .get()
            .await()
        return snapshot.documents.mapNotNull { doc ->
            val symptom = doc.toObject(Symptom::class.java) ?: return@mapNotNull null
            doc.id to symptom
        }
    }

    suspend fun getPoopLogsForDateRange(startDate: Date, endDate: Date): List<Pair<String, PoopLog>> {
        val snapshot = userDoc().collection("poopLogs")
            .whereGreaterThanOrEqualTo("createdAt", Timestamp(startDate))
            .whereLessThan("createdAt", Timestamp(endDate))
            .orderBy("createdAt", Query.Direction.DESCENDING)
            .get()
            .await()
        return snapshot.documents.mapNotNull { doc ->
            val log = doc.toObject(PoopLog::class.java) ?: return@mapNotNull null
            doc.id to log
        }
    }

    suspend fun deleteMeal(mealId: String) {
        userDoc().collection("meals").document(mealId).delete().await()
    }

    suspend fun deleteSymptom(symptomId: String) {
        userDoc().collection("symptoms").document(symptomId).delete().await()
    }

    suspend fun deletePoopLog(logId: String) {
        userDoc().collection("poopLogs").document(logId).delete().await()
    }

    suspend fun deleteCorrelationReport(reportId: String) {
        userDoc().collection("correlationReports").document(reportId).delete().await()
    }

    // --- Helpers ---
    private fun getStartOfDay(): Date {
        val cal = Calendar.getInstance()
        cal.set(Calendar.HOUR_OF_DAY, 0)
        cal.set(Calendar.MINUTE, 0)
        cal.set(Calendar.SECOND, 0)
        cal.set(Calendar.MILLISECOND, 0)
        return cal.time
    }
}

data class RecentEntry(
    val type: String,
    val title: String,
    val subtitle: String,
    val timestamp: Timestamp? = null
)

File: ios/GutHealthPackage/Sources/GutHealthFeature/Services/FirestoreService.swift
import Foundation
import FirebaseFirestore

public actor FirestoreService {
    public static let shared = FirestoreService()
    private let db = Firestore.firestore()

    private init() {}

    // MARK: - User Profile

    public func saveUserProfile(uid: String, profile: UserProfile) throws {
        let ref = db.collection("users").document(uid)
        try ref.setData(from: profile)
    }

    public func getUserProfile(uid: String) async throws -> UserProfile? {
        let doc = try await db.collection("users").document(uid).getDocument()
        guard doc.exists else { return nil }
        var profile = try doc.data(as: UserProfile.self)
        profile.id = doc.documentID
        return profile
    }

    public func updateUserProfile(uid: String, fields: [String: Any]) async throws {
        try await db.collection("users").document(uid).updateData(fields)
    }

    // MARK: - Meals

    public func saveMeal(uid: String, meal: Meal) throws -> String {
        let ref = try db.collection("users").document(uid).collection("meals").addDocument(from: meal)
        return ref.documentID
    }

    public func getMeals(uid: String, limit: Int = 50) async throws -> [Meal] {
        let snapshot = try await db.collection("users").document(uid).collection("meals")
            .order(by: "createdAt", descending: true)
            .limit(to: limit)
            .getDocuments()
        return snapshot.documents.compactMap { doc in
            var meal = try? doc.data(as: Meal.self)
            meal?.id = doc.documentID
            return meal
        }
    }

    // MARK: - Symptoms

    public func saveSymptom(uid: String, symptom: Symptom) throws -> String {
        let ref = try db.collection("users").document(uid).collection("symptoms").addDocument(from: symptom)
        return ref.documentID
    }

    public func getSymptoms(uid: String, limit: Int = 50) async throws -> [Symptom] {
        let snapshot = try await db.collection("users").document(uid).collection("symptoms")
            .order(by: "createdAt", descending: true)
            .limit(to: limit)
            .getDocuments()
        return snapshot.documents.compactMap { doc in
            var symptom = try? doc.data(as: Symptom.self)
            symptom?.id = doc.documentID
            return symptom
        }
    }

    /// Marks a symptom as resolved by setting its endedAt to now.
    public func resolveSymptom(uid: String, symptomId: String, endedAt: Date = Date()) async throws {
        try await db
            .collection("users").document(uid)
            .collection("symptoms").document(symptomId)
            .updateData(["endedAt": Timestamp(date: endedAt)])
    }

    // MARK: - Poop Logs

    public func savePoopLog(uid: String, log: PoopLog) throws -> String {
        let ref = try db.collection("users").document(uid).collection("poopLogs").addDocument(from: log)
        return ref.documentID
    }

    public func getPoopLogs(uid: String, limit: Int = 50) async throws -> [PoopLog] {
        let snapshot = try await db.collection("users").document(uid).collection("poopLogs")
            .order(by: "createdAt", descending: true)
            .limit(to: limit)
            .getDocuments()
        return snapshot.documents.compactMap { doc in
            var log = try? doc.data(as: PoopLog.self)
            log?.id = doc.documentID
            return log
        }
    }

    // MARK: - Photo URL Updates

    public func updateMealPhotoUrl(uid: String, mealId: String, photoUrl: String) async throws {
        try await db.collection("users").document(uid).collection("meals").document(mealId)
            .updateData(["photoUrl": photoUrl])
    }

    public func updatePoopLogPhotoUrl(uid: String, logId: String, photoUrl: String) async throws {
        try await db.collection("users").document(uid).collection("poopLogs").document(logId)
            .updateData(["photoUrl": photoUrl])
    }

    // MARK: - Daily Summaries

    public func getDailySummary(uid: String, date: String) async throws -> DailySummary? {
        let doc = try await db.collection("users").document(uid).collection("dailySummaries").document(date).getDocument()
        guard doc.exists else { return nil }
        var summary = try doc.data(as: DailySummary.self)
        summary.id = doc.documentID
        return summary
    }

    // MARK: - Correlation Reports

    public func getCorrelationReports(uid: String, limit: Int = 20) async throws -> [CorrelationReport] {
        let snapshot = try await db.collection("users").document(uid).collection("correlationReports")
            .order(by: "createdAt", descending: true)
            .limit(to: limit)
            .getDocuments()
        return snapshot.documents.compactMap { doc in
            var report = try? doc.data(as: CorrelationReport.self)
            report?.id = doc.documentID
            return report
        }
    }

    // MARK: - Delete Entries

    public func deleteMeal(uid: String, mealId: String) async throws {
        try await db.collection("users").document(uid).collection("meals").document(mealId).delete()
    }

    public func deleteSymptom(uid: String, symptomId: String) async throws {
        try await db.collection("users").document(uid).collection("symptoms").document(symptomId).delete()
    }

    public func deletePoopLog(uid: String, logId: String) async throws {
        try await db.collection("users").document(uid).collection("poopLogs").document(logId).delete()
    }

    public func deleteCorrelationReport(uid: String, reportId: String) async throws {
        try await db.collection("users").document(uid).collection("correlationReports").document(reportId).delete()
    }

    // MARK: - Meals by Date

    public func getMealsForDate(uid: String, date: Date) async throws -> [Meal] {
        let calendar = Calendar.current
        let startOfDay = calendar.startOfDay(for: date)
        guard let endOfDay = calendar.date(byAdding: .day, value: 1, to: startOfDay) else { return [] }
        let snapshot = try await db.collection("users").document(uid).collection("meals")
            .whereField("createdAt", isGreaterThanOrEqualTo: startOfDay)
            .whereField("createdAt", isLessThan: endOfDay)
            .order(by: "createdAt", descending: true)
            .getDocuments()
        return snapshot.documents.compactMap { doc in
            var meal = try? doc.data(as: Meal.self)
            meal?.id = doc.documentID
            return meal
        }
    }

    public func getSymptomsForDate(uid: String, date: Date) async throws -> [Symptom] {
        let calendar = Calendar.current
        let startOfDay = calendar.startOfDay(for: date)
        guard let endOfDay = calendar.date(byAdding: .day, value: 1, to: startOfDay) else { return [] }
        let snapshot = try await db.collection("users").document(uid).collection("symptoms")
            .whereField("createdAt", isGreaterThanOrEqualTo: startOfDay)
            .whereField("createdAt", isLessThan: endOfDay)
            .order(by: "createdAt", descending: true)
            .getDocuments()
        return snapshot.documents.compactMap { doc in
            var symptom = try? doc.data(as: Symptom.self)
            symptom?.id = doc.documentID
            return symptom
        }
    }

    public func getPoopLogsForDate(uid: String, date: Date) async throws -> [PoopLog] {
        let calendar = Calendar.current
        let startOfDay = calendar.startOfDay(for: date)
        guard let endOfDay = calendar.date(byAdding: .day, value: 1, to: startOfDay) else { return [] }
        let snapshot = try await db.collection("users").document(uid).collection("poopLogs")
            .whereField("createdAt", isGreaterThanOrEqualTo: startOfDay)
            .whereField("createdAt", isLessThan: endOfDay)
            .order(by: "createdAt", descending: true)
            .getDocuments()
        return snapshot.documents.compactMap { doc in
            var log = try? doc.data(as: PoopLog.self)
            log?.id = doc.documentID
            return log
        }
    }

    // MARK: - Weekly Insights

    public func getWeeklyInsights(uid: String, limit: Int = 10) async throws -> [WeeklyInsight] {
        let snapshot = try await db.collection("users").document(uid).collection("weeklyInsights")
            .order(by: "generatedAt", descending: true)
            .limit(to: limit)
            .getDocuments()
        return snapshot.documents.compactMap { doc in
            var insight = try? doc.data(as: WeeklyInsight.self)
            insight?.id = doc.documentID
            return insight
        }
    }

    // MARK: - FODMAP Database

    public func getFodmapFoods() async throws -> [FodmapFood] {
        let snapshot = try await db.collection("fodmapDatabase").getDocuments()
        return snapshot.documents.compactMap { doc in
            var food = try? doc.data(as: FodmapFood.self)
            food?.id = doc.documentID
            return food
        }
    }

    public func searchFodmapFoods(query: String) async throws -> [FodmapFood] {
        let all = try await getFodmapFoods()
        let lowered = query.lowercased()
        return all.filter { $0.name.lowercased().contains(lowered) }
    }

    // MARK: - Delete All User Data

    public func deleteUserData(uid: String) async throws {
        let userRef = db.collection("users").document(uid)
        let subcollections = ["meals", "symptoms", "poopLogs", "dailySummaries", "correlations", "correlationReports", "weeklyInsights"]
        for name in subcollections {
            let snapshot = try await userRef.collection(name).getDocuments()
            for doc in snapshot.documents {
                try await doc.reference.delete()
            }
        }
        try await userRef.delete()
    }
}
